### DISCUSSION TITLE
Synchronize time with AP-client

### DISCUSSION THREAD

**Discussion by tensarflow** [Hieromon/AutoConnect#485] [How can I?] [answered]
Hey there! Thanks for the awesome work!

I am trying to get the correct time to my ESP32 in a scenario where I don't have internet connection. Is it possible to get the time of the client with javascript code? Essentially, a custom webpage with a button "sync time" and when I press the button, the time from my smartphone is being copied to the ESP32.
I'd appreciate it if you could guide me to the right correction.

Thanks

**Comment by Hieromon** [comment, marked as answer]
@tensarflow If you are not getting an Internet connection, then neither configTime nor the POSIX series of the time setting functions is valid. <strike>Which function do you intend to use to set the ESP32 system time? 
If you can show me it,</strike> But It appears that `settimeofday` is valid with ESP-IDF. So, I may be able to suggest an example sketch of what you intend to do.

![timesync](https://user-images.githubusercontent.com/12591771/164259758-bc3e5632-88d5-470b-b931-900a9a1d4318.png)

The client's machine time can be obtained via JavaScript. By forwarding the retrieved the client-time to the AutoConnectInput element, the AutoConnectAux page handler can retrieve its time value. Assuming that the web page has a field for echo back of the client's machine time, client_time. The JSON declaration for that AutoConnectInput would look like this:

```json
{
  "name": "client_time",
  "type": "ACInput",
  "label": "Client Time:",
  "poseterior": "div"
}
```

And the JavaScript code to get the client time is as follows:

```js
var c_clock = document.getElementById('client_time');
const c_time = new Date();
c_clock.value = c_time.toISOString();
```

Next, the value of AutoConnectInput will be used to set the ESP32 time. That is the role of the custom web page handler. It converts the time value as a string to struct tm time information and sets the ESP32 time using `settimeofday` This handler responds to the `SYNC` button as AutoConnectSubmit.

```json
{
  "name": "sync",
  "type": "ACSubmit",
  "value": "SYNC",
  "uri": "/sync"
},
```
```cpp
String handleSync(AutoConnectAux& aux, PageArgument& args) {
  struct tm timeInfo;
  struct timeval  tv;

  // Retrieve the client-time and prepare the struct tm.
  AutoConnectInput& clientTime = pageClock["client_time"].as<AutoConnectInput>();
  Serial.printf("clientTime:%s\n", clientTime.value.c_str());
  strptime(clientTime.value.c_str(), "%Y-%m-%dT%TZ", &timeInfo);

  // It allows you to confirm the time that should be set to the ESP32 with Serial.
  char timeString[20];
  strftime(timeString, sizeof(timeString), "%F %T", &timeInfo);
  Serial.printf("timeString:%s\n", timeString);

  // Set ESP32 time
  time_t  t = mktime(&timeInfo);
  tv.tv_sec = t;
  tv.tv_usec = 0;
  settimeofday(&tv, NULL);

  // Redirects the /timesync page without returning a response as page content.
  aux.redirect(ESPTIME_TIME_URL);
  return String();
}
```

The full sketch code for the custom web page with AutoConnect shown at the top of this answer looks like this:
You can try it by navigating to ESP32's `local-IP/timesync` from your client browser.

```cpp
#include <Arduino.h>
#include <WiFi.h>
#include <WebServer.h>
#include <time.h>
#include <AutoConnect.h>

#define ESPTIME_TIME_URL  "/timesync"
#define ESPTIME_EP_URL  "/esp_time"

static const char PAGE_TIMECLOCK[] = R"(
{
  "uri": ")" ESPTIME_TIME_URL R"(",
  "title": "Sync",
  "menu": true,
  "element": [
    {
      "name": "client_time",
      "type": "ACInput",
      "label": "Client Time:",
      "poseterior": "div"
    },
    {
      "name": "esp_time",
      "type": "ACElement",
      "value": "<div>ESP Time: <span id='esp-time'></span></div>"
    },
    {
      "name": "sync",
      "type": "ACSubmit",
      "value": "SYNC",
      "uri": "/sync"
    },
    {
      "name": "c_clock",
      "type": "ACElement",
      "value": "<script>window.onload=()=>{var c_clock=document.getElementById('client_time');var esp_clock=document.getElementById('esp-time');var busy=false;setInterval(()=>{const c_time=new Date();c_clock.value=c_time.toISOString();if(!busy){busy=true;fetch(')" ESPTIME_EP_URL R"(').then(response=>{return response.text();}).then(text=>{esp_clock.innerText=text;busy=false;}).catch(error=>{esp_clock.innerText=error;busy=false;});}},1000);};</script>"
    }
  ]
}
)";

static const char PAGE_TIMESYNC[] = R"(
{
  "uri": "/sync",
  "title": "Sync",
  "menu": false,
  "response": false
}
)";

WebServer   server;
AutoConnect portal(server);
AutoConnectAux pageClock;
AutoConnectAux pageSync;

String handleSync(AutoConnectAux& aux, PageArgument& args) {
  struct tm timeInfo;
  struct timeval  tv;

  // Retrieve the client-time and prepare the struct tm.
  AutoConnectInput& clientTime = pageClock["client_time"].as<AutoConnectInput>();
  Serial.printf("clientTime:%s\n", clientTime.value.c_str());
  strptime(clientTime.value.c_str(), "%Y-%m-%dT%TZ", &timeInfo);

  // It allows you to confirm the time that should be set to the ESP32 with Serial.
  char timeString[20];
  strftime(timeString, sizeof(timeString), "%F %T", &timeInfo);
  Serial.printf("timeString:%s\n", timeString);

  // Set ESP32 time
  time_t  t = mktime(&timeInfo);
  tv.tv_sec = t;
  tv.tv_usec = 0;
  settimeofday(&tv, NULL);

  // Redirects the /timesync page without returning a response as page content.
  aux.redirect(ESPTIME_TIME_URL);
  return String();
}

void onGetESPTime() {
  struct tm timeInfo;
  char  timeString[20];

  getLocalTime(&timeInfo);
  strftime(timeString, sizeof(timeString), "%F %T", &timeInfo);
  server.send(200, "text/plain", String(timeString));
}

void setup() {
  Serial.begin(115200);
  Serial.println();

  // Load clock page
  pageClock.load(PAGE_TIMECLOCK);

  // Load sync page and register the handler
  pageSync.load(PAGE_TIMESYNC);
  pageSync.on(handleSync);

  portal.join({ pageClock, pageSync });
  portal.begin();

  server.on(ESPTIME_EP_URL, onGetESPTime);
}

void loop() {
  portal.handleClient();
}
```

**Comment by Hieromon** [comment]
@tensarflow Considering that clients connecting to the ESP module generally hold adjusted times, I can derive an enhancement to AutoConnect from your question. All of those enhancements are the ability to reflect client time into the ESP module. The time set by this is not directly derived from NTP, but an approximate value can still be expected.

1. Incorporate the Sync Time item into the AutoConenct menu.
2. Supports new AutoConnectElement like as AutoConnectTimeSync. It includes necessary JavaScript and the server side reponder logic. (It will set to client-time to ESP module)

Are these features beneficial to many users? I don't want to add redundant elements, so I'd like to keep it as simple as possible.

**Reply by tensarflow** [reply]
Thank you for the answer. You are a hero :)
I will implement this and give you feedback.

Regarding integration, I would definitely benefit from it :), but generally also looks like something reasonable.

**Comment by tensarflow** [comment]
Hello again!
I tried the code. It works as expected. One minor issue is the time zone. By synchronizing the time via a client I actually was also trying to overcome the time zone setting problem. On the JavaScript side you use the `toISOString()` function which always gives the time as 0 UTC. How would you tackle this?
Thanks in advance

**Reply by tensarflow** [reply]
I managed to get the correct time using `toLocaleString()` instead of `toISOString` and parsing the produced string like so:

```C++
strptime(clientTime.value.c_str(), "%d.%m.%Y, %H:%M:%S", &timeInfo);
```

Thanks again for your help!

**Reply by tensarflow** [reply]
BTW I noticed that strptime compiles correctly but my IDE -VS Code + PlatformIO- doesnt recognize it and gives the undefined error. Would you happen to know why that is?
